### DISCUSSION TITLE
sclr.ac - Scaler's Official URL Shortener

### DISCUSSION THREAD

**Discussion by championswimmer** [scaleracademy/open-source-projects#3] [Projects]
# sclr.ac 
sclr.ac is Scaler's official URL shortener - it will be used to create shortlinks for masterclasses, events, referral links etc. 

## Details 
Repositories 
- [backend](https://github.com/scaleracademy/sclr.ac-backend)
	- [alt new backend](https://github.com/championswimmer/URL_Shortner_Discord_Session) 
- [frontend](https://github.com/scaleracademy/sclr.ac-frontend)

Maintainers 
- [Arnav Gupta](https://github.com/championswimmer)

## Summary 
The URL shorten-er project - sclr.ac works like common URL shortening services like http://bit.ly or http://goo.gl etc.

People can -

- Create shortcodes for their URLs (which are randomly generated 5 digit alphanumeric)
- Create custom shortcodes (like /mycode) for their URLs

There would be stats available for how many times each shortlink has been opened. Also the traffic stats will have details of the referrer URL (which gives information on where the traffic is coming from).

### Features (future) 

- Can links expire ? 
  	- Expire if `hits < X && created more than Y days ago` (eg: expire unused links after 1 month) 
- Subscriptions 
	- Free account = 5 short codes
	- Lite Account = 20 short codes / month 
	- Pro Account = Unlimited short codes 

#### Base Entity 
Every entity extends from this (i.e. all tables have these columns) 
- `id` : integer autoincrement
- `createdAt`: date
- `updatedAt`: date 

#### URLs 
- `id`: bigint (basically base64 representation of shortCode) 
- `longUrl`: string - the actual HTTP url which we are shortening 
- `shortCode`: string - the shortened string 
- `ownerId`: id of user who created it 
- `hits`: big int 

eg: if https://medium.com/@championswimmer is shortened to https://sclr.ac/agm then `agm` is the `shortCode` 

#### Visits 
- `urlId`: id of shortcode which was visited  
- `ipAddr`: string - IP address of visiting user 
- `referrer`: string - the previous website someone was visiting 

#### Users 
- `username`: string
- `email`: string   
- `auth`: ??? 

### API Endpoints 
| Legend | Meaning | 
| --------| ---------|
| ⚠️ | Temporary, not for prod | 
| 🔐 | Needs authentication |
| 👮 | Admins are allowed |
| 🧑| Only users' own info (if they have created that entity) |
| 📄 | Has pagination |

##### `POST /users/` ⚠️
Register a new user with email/password 

##### `POST /users/login` ⚠️
Login a new user with email/password 

##### `GET /users/login`
Redirect to Google for Google Oauth 

##### `GET /users/login/callback`
Callback from Google for Oauth  

##### `POST /urls/` 🔐
Create a new shortened URL 

For getting random short code 
``` 
{
	"longUrl": "https://example.com/mysite/?some=query"
}
```

For getting custom short code 
``` 
{
	"longUrl": "https://example.com/mysite/?some=query",
	"shortCode": "mylink" 
}
```

##### `GET /urls` 🔐 📄
Shows all short URLs
- For admins, shows **ALL** urls on our system 
- For normal users shows only urls _they have created_ 

```
GET https://sclr.ac/urls?contains=google 

Should show all URLs that have "google" in their longUrl 
```

##### `GET /urls/{shortCode}` 🔐👮🧑
Get details of a short code (like its long URL, no of hits etc)

##### `PATCH /urls/{shortCode}` 🔐👮🧑
Update details of a shortCode (for now mainly to update the long URL) 

##### `GET /{shortCode}` 
‼️ This is not really an API endpoint
This will redirect the user to the longUrl. People will use this endpoint directly from the browser 

##### `GET /users/me` 🔐
See my profile details 

##### `PATCH /users/me` 
Update my profile details (if any) 

### Shortening Logic 

Given long URL: `https://github.com/scaleracademy/open-source-projects/discussions/3` 
We need a short code: `https://sclr.ac/gbv6j` i.e. `gbv6j` 

Shortcode requirements: 
- Contains `A-Z`, `a-z`, `0-9` `_` and `-` 
- Length: **8**
- DB data type: VARCHAR(8) 

`0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_-` 
64^8 == 2^X

Take `gbv6j` as a base64 representation, convert it to an integer and we'll use that as the `id` of our URLs table.

**Comment by harikanani** [comment]
Do we need link expiration fetures or not?

**Comment by kunwar97** [comment]
@championswimmer I have some ideas:

1. Add Authentication & My Urls module
2. LRU Caching on backend for less latency

Let me know if i can start working on these features
